File: frontend/test/utils/study_calendar_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:srbski_read/utils/study_calendar.dart';

void main() {
  test('месяц следует серверной дате и учитывает високосный год', () {
    final month = studyMonth('2024-02-28')!;
    expect(month.dayCount, 29);
    expect(month.padding, 3);
    expect(month.date(29), '2024-02-29');
    expect(studyMonth('2026-01-01', -1)!.date(1), '2025-12-01');
    expect(studyMonth('2026-02-31'), null);
  });
  test('старый кеш не заменяет новый снимок статистики', () {
    final initial = {'asOf': '2026-09-09T12:00:00Z', 'current': 5};
    final old = {'asOf': '2026-09-08T12:00:00Z', 'current': 4};
    expect(latestStudyData(initial, old), initial);
    expect(latestStudyData(null, old), old);
  });
}

File: web/src/lib/studyCalendar.test.ts
import { expect, it } from 'vitest';
import { latestStudy, studyMonth } from './studyCalendar';
import type { Study } from '../api/personal';

it('расставляет даты с понедельника, включая високосный февраль и границу года', () => {
  expect(studyMonth('2024-03-01', -1)).toMatchObject({ key: '2024-02', padding: 3 });
  expect(studyMonth('2024-03-01', -1)?.days).toHaveLength(29);
  expect(studyMonth('2026-01-01', -1)?.key).toBe('2025-12');
  expect(studyMonth('2026-09-09')?.days[8]).toBe('2026-09-09');
  expect(studyMonth('broken')).toBeNull();
});

it('старый кеш не отменяет более свежую статистику профиля', () => {
  const initial = { current: 3, asOf: '2026-09-09T12:00:00Z' } as Study;
  const live = { current: 1, asOf: '2026-09-09T11:00:00Z' } as Study;
  expect(latestStudy(initial, live)?.current).toBe(3);
  expect(latestStudy(initial, { ...live, current: 4, asOf: '2026-09-09T13:00:00Z' })?.current).toBe(4);
});
